src/dataworker/Dataworker.ts

```diff
@@ -1759,94 +1760,107 @@ export class Dataworker {
-            message: `Skipping exchange rate update for ${tokenSymbol} because current liquid reserves > netSendAmount for chain ${leaf.chainId}`,
-            l2ChainId: leaf.chainId,
-            currHubPoolLiquidReserves,
-            netSendAmount: leaf.netSendAmounts[idx],
-            l1Token,
-          });
-          return;
-        }
-
-        // @dev: Virtual balance = post-sync liquid reserves + any used balance.
-        const multicallInput = [
-          hubPool.interface.encodeFunctionData("sync", [l1Token]),
-          hubPool.interface.encodeFunctionData("pooledTokens", [l1Token]),
-        ];
-        const multicallOutput = await hubPool.callStatic.multicall(multicallInput);
-        const updatedPooledTokens = hubPool.interface.decodeFunctionResult("pooledTokens", multicallOutput[1]);
-        const updatedLiquidReserves = updatedPooledTokens.liquidReserves;
-        const virtualHubPoolBalance = updatedLiquidReserves.sub(
-          balanceAllocator.getUsed(hubPoolChainId, l1Token, hubPool.address)
-        );
+        aggregateNetSendAmounts[l1Token] = aggregateNetSendAmounts[l1Token].add(leaf.netSendAmounts[idx]);
+      });
+    });
 
-        // If the virtual balance is still too low to execute the pool leaf, then log an error that this will
-        // pool rebalance leaf execution will fail.
-        if (virtualHubPoolBalance.lt(leaf.netSendAmounts[idx])) {
-          this.logger.error({
-            at: "Dataworker#executePoolRebalanceLeaves",
-            message: "Executing pool rebalance leaf on HubPool will fail due to lack of funds to send.",
-            leaf: leaf,
-            l1Token,
-            netSendAmount: leaf.netSendAmounts[idx],
-            updatedLiquidReserves,
-            virtualHubPoolBalance,
-          });
-          return;
-        }
+    // Now, go through each L1 token and see if we need to update the exchange rate for it.
+    await sdkUtils.forEachAsync(Object.keys(aggregateNetSendAmounts), async (l1Token) => {
+      const requiredNetSendAmountForL1Token = aggregateNetSendAmounts[l1Token];
+      // If netSendAmounts is 0, there is no need to update this exchange rate.
+      assert(requiredNetSendAmountForL1Token.gte(0), "Aggregate net send amount should be >= 0");
+      if (requiredNetSendAmountForL1Token.eq(0)) {
+        return;
+      }
+      // The "used" balance kept in the BalanceAllocator should have adjusted for the netSendAmounts and relayer refund leaf
+      // executions above. Therefore, check if the current liquidReserves is less than the pool rebalance leaf's netSendAmount
+      // and the virtual hubPoolBalance would be enough to execute it. If so, then add an update exchange rate call to make sure that
+      // the HubPool becomes "aware" of its inflow following the relayre refund leaf execution.
+      const currHubPoolLiquidReserves =
+        latestLiquidReserves[l1Token] ?? this.clients.hubPoolClient.getLpTokenInfoForL1Token(l1Token).liquidReserves;
```

Comment:
`HubPoolClient.getLpTokenInfoForL1Token()` can return `undefined`. Dereferencing `liquidReserves`  will cause a crash in that case, whereas an assert will at least make it clear what happened.

```suggestion
        latestLiquidReserves[l1Token] ?? this.clients.hubPoolClient.getLpTokenInfoForL1Token(l1Token)?.liquidReserves;
```